tests/test_rr_models.py

```diff
@@ -0,0 +1,30 @@
+import ast
+
+from sqlalchemy import create_mock_engine
+from sqlalchemy.sql.type_api import TypeEngine
+
+from ukrr_models.rr_models import Base
+from tests.table_extractor import TableNameExtractor
+
+
+def mssql_dump(sql: TypeEngine, *args, **kwargs):
+    dialect = sql.compile(dialect=mssql_engine.dialect)
+    if sql_str := str(dialect).rstrip():
+        print(f"{sql_str};")
+
+
+mssql_engine = create_mock_engine("mssql+pyodbc://", mssql_dump)
+
+with open("./ukrr_models/rr_models.py", "r") as file:
+    code = file.read()
+
+tree = ast.parse(code)
+table_name_extractor = TableNameExtractor()
+table_name_extractor.visit(tree)
+
+
+def test_create_tables(capsys):
```

Comment:
This seems nifty, but is it effectively testing that sqlalchemy successully generates DDL for each of the models - as much of a test of sqlalchemy than of our models?

There are copies of the UKRR DDL at https://github.com/renalreg/sql-backup/tree/master/backup/RR-SQL-LIVE/RENALREG/TABLES which you could perhaps compare with whatever SQLA is generating or do something like https://stackoverflow.com/questions/30428639/check-database-schema-matches-sqlalchemy-models-on-application-startup

Reply:
This parses the models file looking for __tablename__ and just compares that will the DDL. It is desgined to catch any issues in the written code that would prevent the tables from building (like specifying incorrect types on FK columns). You could achieve the same thing by just doing create_all() and looking for errors but I found some case where a table silently failed to build.

I see this as a basic check on models but taking a more holistic approach and making sure what we have actually mirrors what is going on in the database might not be a bad idea. Are those DDLs updated regularly and automatically?